### ISSUE TITLE
functools.cached_property and derived classes

### ISSUE THREAD

**Issue by scole66** [python/typeshed#3547]
`functools.cached_property` doesn't seem to create good annotations for derived classes. Is this a typeshed thing or a mypy thing? 

Testcase:
```python
from functools import cached_property
class baseclass:
    def __init__(self, data: str):
        self.data = data
    @cached_property
    def characteristic(self) -> int:
        return len(self.data)
class derived(baseclass):
    @cached_property
    def characteristic(self) -> int:
        return 10
```
Output from mypy 0.760:
```
$ mypy mypytest.py 
mypytest.py:10: error: Signature of "characteristic" incompatible with supertype "baseclass"
Found 1 error in 1 file (checked 1 source file)
```

**Comment by JelleZijlstra** [comment]
For what it's worth the typeshed definition is:
```
    class cached_property(Generic[_S, _T]):
        func: Callable[[_S], _T]
        attrname: Optional[str]
        def __init__(self, func: Callable[[_S], _T]) -> None: ...
        @overload
        def __get__(self, instance: None, owner: Optional[Type[_S]] = ...) -> cached_property[_S, _T]: ...
        @overload
        def __get__(self, instance: _S, owner: Optional[Type[_S]] = ...) -> _T: ...
        def __set_name__(self, owner: Type[_S], name: str) -> None: ...
```
It's not immediately obvious to me why mypy is giving the errors you mention—maybe something to do with the invariant TypeVars that `cached_property` is generic over? It may be helpful to `reveal_type(characteristic)` inside the class body.

**Comment by scole66** [comment]
```
mypytest.py:12: note: Revealed type is 'functools.cached_property[mypytest.baseclass*, builtins.int*]'
mypytest.py:18: note: Revealed type is 'functools.cached_property[mypytest.derived*, builtins.int*]'
```

It's the same outside of the classes as well.

**Comment by scole66** [comment]
The revealed types _seem_ accurate; the problem seems to be that there's an issue with the link between `baseclass` and `derived`. Which makes this seem like a mypy issue.

But `@property` works fine, though the typeshed definition looks completely different to me, which makes this seem like a typeshed issue.

**Comment by JelleZijlstra** [comment]
`@property` is heavily special-cased in mypy itself, so its typeshed implementation doesn't necessarily tell you much.

I realized the issue here is the first type argument in the stub. If I remove that argument and instead use `Any`, the error goes away. I also tried making this typevar covariant instead, but that leads to `error: Cannot use a covariant type variable as a parameter` and doesn't fix the "incompatible with supertype" error. I'll submit a PR for the `Any` version.
